### DISCUSSION TITLE
[v3] Has anyone figured out how to catch and handle custom exceptions?

### DISCUSSION THREAD

**Discussion by vanere** [livewire/livewire#5976] [1. Help]
In version 2 we could intercept and handle custom exceptions thrown by overriding the syncInput method of Livewire's HandlesActions trait.  See below...

```PHP
   use HandlesActions {
        syncInput as protected livewireSyncInput;
    }

    public function syncInput($name, $value, $rehash = true): void
    {
        try {
            $this->livewireSyncInput($name, $value, $rehash);
        }
        catch (InvalidValueObjectValueException $exception) {
            $this->getErrorBag()->add($name, $exception->getMessage());
        }
    }
```

I have not been able to figure out how this can be accomplished in V3 yet so I was wondering if anyone else has been able to accomplish this.

The weird edge case is that self validating value objects are used in the app and they throw a specific exception when a value is invalid. Its a nice way of centralizing validation (in this app). 

When those validation exceptions are thrown they display the exception modal. See below...

![Screen Shot 2023-07-25 at 9 07 18 PM](https://github.com/livewire/livewire/assets/3731011/bfc106f8-9586-4e22-85b5-747745375a1d)

If anyone else has been able to capture custom exceptions in v3 please let me know.

**Comment by joshhanley** [comment]
@vanere are you able to provide some more details on why you needed to do that in V2 and what you're hoping to achieve by it in V3, so I can then advise the best way to do it? Thanks!

**Reply by vanere** [reply]
Hey, yes absolutely.

When using Livewire V2's [Binding Directly To Model Properties](https://laravel-livewire.com/docs/2.x/properties#binding-models) and the property on the model was cast to a value object using Laravel's [Value Object Casting](https://laravel.com/docs/10.x/eloquent-mutators#value-object-casting), if an exception is thrown in the caster class or the value object while it's being created (maybe due to a value being invalid) then that exception is propagated up to Livewire which displays a modal with the stack trace and other info about the state of your application at the time of the exception.

One example is a birth_date field on a User class that is cast to a BirthDate value object using Laravel's Value Object Casting. The object it is cast to self validates the value. So if across the application if a birth date is needed, instead of passing a Date or Carbon instance directly (which can contain dates in the future and therefore are not valid birth dates), the BirthDate object is passed and this guarantees that the variable is always a valid birth date. 

Livewire v2 did not exactly play well with custom exceptions thrown in value object instantiations or custom exceptions thrown in the casting of eloquent properties to value object and displayed the exception modal that pops up when non validation exceptions are thrown. In order to catch these exceptions and display them in the UI, the "syncInput" method in the "HandlesActions" trait was overwritten as shown above. This allowed for the exception message to be displayed for the correct form input by adding the exception message to the error bag with the property's key (eg. 'user.birth_date').

In V3 I have tried to use both Synthesizers and Wireables and they work as expected when a valid value is being used, however when invalid values are used and an exception is thrown while creating the value object, Livewire shows the exception modal with the stack trace and I have no been able to find a way to intercept that exception as I did in v2 by overriding the syncInput method of the HandlesActions trait on the Livewire\Component class. 

Any help or guidance would be appreciated. And if you have more questions please let me know. (sorry for any typos, I typed this while in a rush)

**Reply by joshhanley** [reply]
@vanere thanks! Hmm the new synths should fix that. Can chance you could share your code of what you tried with the synths and an example component class and view so we can investigate? Thanks!
